### PR TITLE
Feature/32 dataframe framework

### PR DESCRIPTION
**Pull request by nils-schmitt**
Working implementation of the dataframe framework.

Please note, that output_handling is removed due to handling issues of dataframe / the csv file.

### CODE REVIEW

File: tracex/extraction/prototype/utils.py [hunk 1]
```diff
@@ -41,11 +43,31 @@ def get_decision(question):
...
 
-def query_gpt(messages, temperature=TEMPERATURE_SUMMARIZING):
+# def query_gpt(messages, temperature=TEMPERATURE_SUMMARIZING):
+#     """Queries the GPT engine."""
+#     response = client.chat.completions.create(model=MODEL,
+#     messages=messages,
+#     max_tokens=MAX_TOKENS,
+#     temperature=temperature)
+#     output = response.choices[0].message.content
+#     return output
```
Comment (FR-SON) [inline tracex/extraction/prototype/utils.py:53]: Is this still required? Otherwise, please remove.

**Review by FR-SON — CHANGES_REQUESTED** [review]
Looks very good overall. I didn't run it with the changes though, only read through the changes. There are some minor improvements that should be made. 
It might be beneficial to document the insights you had that led to the changed prompts, as we will probably work on that topic further in the future.

File: tracex/extraction/prototype/input_handling.py [hunk 2]
```diff
@@ -1,215 +1,363 @@
...
-    with open(
-        (u.output_path / "intermediates/3_bulletpoints_with_end.txt"),
-        "w",
-    ) as f:
-        f.write(bulletpoints_start)
-    return bulletpoints_start
+    name = "end_date"
+    new_df = pd.DataFrame([], columns=[name])
+    values_list = df.values.tolist()
+    i = 0
```
[the remark this answers is not in the document]
Comment (nils-schmitt) [reply tracex/extraction/prototype/input_handling.py]: The i was solely for debugging purposes, it is now removed completely

File: tracex/extraction/prototype/input_handling.py [hunk 3]
```diff
@@ -1,215 +1,299 @@
...
+            if row != "\n":
+                content.append(row)
+        content = content[1:]
+    with open(u.CSV_ALL_TRACES, "a") as f:
+        for row in content:
+            row = row.replace(row[0], str(int(row[0]) + trace_count + 1), 1)
+            f.writelines(row)
+
+
+def farewell():
```
Comment (PitButtchereit) [inline tracex/extraction/prototype/input_handling.py:277]: This is not necessary

File: tracex/extraction/prototype/input_handling.py [hunk 4]
```diff
@@ -1,215 +1,299 @@
...
+        {"role": "assistant", "content": p.TEXT_TO_EVENTINFORMATION_ANSWER},
     ]
     bulletpoints = u.query_gpt(messages)
-    bulletpoints = remove_commas(bulletpoints)
-    bulletpoints = add_ending_commas(bulletpoints)
-    with open((u.output_path / "intermediates/1_bulletpoints.txt"), "w") as f:
-        f.write(bulletpoints)
-    return bulletpoints
+    df = pd.DataFrame([], columns=["event_information"])
+    bulletpoints = bulletpoints.replace("- ", "")
```
Comment (PitButtchereit) [inline tracex/extraction/prototype/input_handling.py:60]: Have you tried to specify in the prompt that the activity labels should be in a specific output format? Then we would not have to do this formatting stuff.

File: tracex/extraction/prototype/prompts.py [hunk 5]
```diff
@@ -53,14 +55,18 @@ def get_date():
             "content": "Please give me one date between 01/01/2020 and 01/09/2023.",
         }
     ]
-    country = client.chat.completions.create(model=u.MODEL, messages=message, max_tokens=50, temperature=0.5)
+    country = client.chat.completions.create(
```
Comment (PitButtchereit) [inline tracex/extraction/prototype/prompts.py:58]: this variable should not be named "country" as this is the function to get a random date

File: tracex/extraction/prototype/prompts.py [hunk 6]
```diff
@@ -115,126 +120,149 @@ def life_circumstances_prompt(sex):
...
 
-# Adding of a start date to every bullet point
-BULLETPOINTS_START_DATE_CONTEXT = """
-    You are an expert in text understanding and your job is to take a given text and given summarizing bulletpoints and to add a start date to every bulletpoint.
-    Edit the bulletpoints in a way, that you just take the existing bulletpoints and add a start date at the end of it.
-    The information about the start date should be extracted from the text or from the context and should be as precise as possible.
-    Do not modify the content of the bulletpoint and keep ending commas.
+# Adding of a start date to every bulletpoint
+START_DATE_CONTEXT = """
+    You are an expert in text understanding and your job is to take a given text and a given bulletpoint and to extract a start date to this bulletpoint.
```
Comment (PitButtchereit) [inline tracex/extraction/prototype/prompts.py:125]: Please change the word bulletpoint to something else like event information

File: tracex/extraction/prototype/prompts.py [hunk 7]
```diff
@@ -115,126 +120,149 @@ def life_circumstances_prompt(sex):
...
     If there is really no information about the start date to be extracted from the text but there is information about events happening at the same time,
     use that information to draw conclusions about the start dates.
-    If there is no information about the start date at all and there is no way of finding some, delete that bulletpoint.
-    The only output should be the updated bullet points, nothing else!
+    If there is only a month specified, use the first of this month as start date. If there is no date specified in the text conclude 'N/A'.
 """
-BULLETPOINTS_START_DATE_PROMPT = """
-    Here is the text and the bulletpoints for which you should extract start dates:
+START_DATE_PROMPT = """
+    Here is the text and the bulletpoint for which you should extract the start date in the format YYYYMMDD with the postfix T0000!
```
Comment (PitButtchereit) [inline tracex/extraction/prototype/prompts.py:136]: Same as above

File: tracex/extraction/prototype/prompts.py
Comment (PitButtchereit) [inline tracex/extraction/prototype/prompts.py:1]: In many prompts we still use the word "bulletpoint". Would be nice to change this

File: tracex/extraction/prototype/input_handling.py [hunk 8]
```diff
@@ -1,215 +1,299 @@
...
+        )
+        new_row = pd.DataFrame([end_date], columns=[name])
+        new_df = pd.concat([new_df, new_row], ignore_index=True)
+        document_intermediates(output)
+    df = pd.concat([df, new_df], axis=1)
+    return df
 
 
-def add_event_types(bulletpoints_durations):
+def add_durations(df):
```
Comment (PitButtchereit) [inline tracex/extraction/prototype/input_handling.py:146]: I thought we would leave out durations for now

**Review by PitButtchereit — CHANGES_REQUESTED** [review]
Please implement some minor changes.

Also I am not sure about the idea to write the dataframe to a file after each pipeline step. What was your motivation about this?

**Review by PitButtchereit — APPROVED** [review]
Joop Joop

**Review by thangixd — APPROVED** [review]
looks good to me. Good work

**Review by FR-SON — APPROVED** [review]

**Pull request merged**